// typedoc-plugin-markdown/dist/theme/context/partials/comments.comment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.comment = comment;
const markdown_1 = require("../../../libs/markdown");
const utils_1 = require("../../../libs/utils");
function comment(model, options = {}) {
    const opts = {
        headingLevel: undefined,
        showSummary: true,
        showTags: true,
        showReturns: false,
        isTableColumn: false,
        ...options,
    };
    const md = [];
    //Add flags and summary
    if (opts.showSummary) {
        // Add flags
        const flagsNotRendered = [
            '@showCategories',
            '@showGroups',
            '@hideCategories',
            '@hideGroups',
        ];
        const flags = [];
        for (const tag of model.modifierTags) {
            if (!flagsNotRendered.includes(tag)) {
                flags.push((0, markdown_1.bold)((0, markdown_1.backTicks)((0, utils_1.camelToTitleCase)(tag.substring(1)))));
            }
        }
        md.push(flags.join(' '));
        // Add summary
        if (model.summary?.length > 0) {
            md.push(this.helpers.getCommentParts(model.summary));
        }
    }
    const blockTagsPreserveOrder = this.options.getValue('blockTagsPreserveOrder');
    const showTags = opts.showTags || (opts.showSummary && blockTagsPreserveOrder.length > 0);
    // Add Tags
    if (showTags && model.blockTags?.length > 0) {
        const blockTags = model.blockTags.reduce((previous, current) => {
            if (current.tag === '@example') {
                const prevExampleTag = previous.find((tag) => ['@example', '@examples'].includes(tag.tag));
                if (prevExampleTag) {
                    return previous.map((prevTag) => {
                        if (prevTag === prevExampleTag) {
                            current.content.unshift({ kind: 'text', text: '\n\n' });
                            return {
                                ...prevTag,
                                tag: '@examples',
                                content: [...prevTag.content, ...current.content],
                            };
                        }
                        return prevTag;
                    });
                }
            }
            return [...previous, current];
        }, []);
        const filteredBlockTags = opts.showReturns ? [] : ['@returns'];
        const tags = blockTags
            .filter((tag) => !filteredBlockTags.includes(tag.tag))
            .filter((tag) => {
            if (!opts.isTableColumn &&
                !(opts.showSummary && opts.showTags) &&
                blockTagsPreserveOrder.length) {
                return opts.showSummary
                    ? blockTagsPreserveOrder.includes(tag.tag)
                    : !blockTagsPreserveOrder.includes(tag.tag);
            }
            return true;
        })
            .filter((tag) => !opts.isTableColumn ||
            (opts.isTableColumn && tag.tag !== '@defaultValue'))
            .map((tag) => {
            const tagText = this.internationalization.translateTagName(tag.tag);
            const tagMd = [
                opts.headingLevel
                    ? (0, markdown_1.heading)(opts.headingLevel, tagText) + '\n'
                    : (0, markdown_1.bold)(tagText) + '\n',
            ];
            tagMd.push(this.helpers.getCommentParts(tag.content));
            return tagMd.join('\n');
        });
        md.push(tags.join('\n\n'));
    }
    const output = md.join('\n\n');
    const parsedOutput = this.options.getValue('sanitizeComments')
        ? (0, utils_1.sanitizeComments)(output)
        : output;
    return parsedOutput;
}
