// typedoc-plugin-markdown/dist/renderer/overrides.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDocs = generateDocs;
exports.render = render;
const events_1 = require("../events");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const typedoc_1 = require("typedoc");
/**
 * Replacement of TypeDoc's Application.generateDocs method to decouple HTML logic.
 */
async function generateDocs(project, out) {
    const start = Date.now();
    await this.renderer.render(project, out);
    if (this.logger.hasErrors()) {
        this.logger.error(this.i18n.docs_could_not_be_generated());
    }
    else {
        this.logger.info(this.i18n.docs_generated_at_0(nicePath(out)));
        this.logger.verbose(`Markdown rendering took ${Date.now() - start}ms`);
    }
}
/**
 * Replacement of TypeDoc's Renderer.render method to decouple HTML logic.
 *
 * This is essentially a copy of the base method with a few tweaks.
 *
 * - Removes unnecessary async calls to load highlighters only required for html theme.
 * - Removes hooks logic that are jsx specific.
 * - Adds any logic specific to markdown rendering.
 */
async function render(project, outputDirectory) {
    this.renderStartTime = Date.now();
    if (this.cleanOutputDir) {
        try {
            fs.rmSync(outputDirectory, { recursive: true, force: true });
        }
        catch {
            this.application.logger.warn('Could not empty the output directory.');
            return;
        }
    }
    try {
        fs.mkdirSync(outputDirectory, { recursive: true });
    }
    catch {
        this.application.logger.error(`Could not create output directory ${outputDirectory}.`);
        return;
    }
    if (this.application.options.isSet('githubPages') &&
        this.application.options.getValue('githubPages') === true) {
        try {
            fs.writeFileSync(path.join(outputDirectory, '.nojekyll'), '');
        }
        catch {
            this.application.logger.warn(this.application.i18n.could_not_write_0(path.join(outputDirectory, '.nojekyll')));
        }
    }
    this.prepareTheme();
    const output = new events_1.MarkdownRendererEvent(events_1.MarkdownRendererEvent.BEGIN, outputDirectory, project);
    output.urls = this.theme.getUrls(project);
    output.navigation = this.theme.getNavigation(project);
    this.trigger(output);
    await Promise.all(this.preRenderAsyncJobs.map((job) => job(output)));
    this.preRenderAsyncJobs = [];
    this.trigger(events_1.MarkdownRendererEvent.BEGIN, output);
    this.application.logger.verbose(`There are ${output.urls?.length} pages to write.`);
    output.urls
        ?.filter((urlMapping) => urlMapping.model instanceof typedoc_1.Reflection)
        .forEach(async (urlMapping) => {
        const [template, page] = output.createPageEvent(urlMapping);
        page.contents = '';
        this.trigger(events_1.MarkdownPageEvent.BEGIN, page);
        if (page.model instanceof typedoc_1.Reflection) {
            page.contents = page.contents + this.theme.render(page, template);
        }
        else {
            throw new Error('Should be unreachable');
        }
        this.trigger(events_1.MarkdownPageEvent.END, page);
        try {
            writeFileSync(page.filename, page.contents);
        }
        catch {
            this.application.logger.error(this.application.i18n.could_not_write_0(page.filename));
        }
    });
    // copy resolved files to media directory
    const media = path.join(outputDirectory, '_media');
    const toCopy = project.files.getNameToAbsoluteMap();
    for (const [fileName, absolute] of toCopy.entries()) {
        copySync(absolute, path.join(media, fileName));
    }
    // process postRenderAsyncJobs
    await Promise.all(this.postRenderAsyncJobs.map((job) => job(output)));
    this.postRenderAsyncJobs = [];
    this.trigger(events_1.MarkdownRendererEvent.END, output);
    this.theme = void 0;
}
/**
 * Some useful utility functions - essentially cherry picked from:
 *
 * - https://github.com/TypeStrong/typedoc/blob/master/src/lib/utils/fs.ts
 * - https://github.com/TypeStrong/typedoc/blob/master/src/lib/utils/path.ts
 */
/**
 *  Writes a file to disc.
 */
function writeFileSync(fileName, data) {
    fs.mkdirSync(path.dirname(normalizePath(fileName)), { recursive: true });
    fs.writeFileSync(normalizePath(fileName), data);
}
/**
 * Returns a readable path from an absolute path.
 */
function nicePath(absPath) {
    if (!path.isAbsolute(absPath))
        return absPath;
    const relativePath = path.relative(process.cwd(), absPath);
    if (relativePath.startsWith('..')) {
        return normalizePath(absPath);
    }
    return `./${normalizePath(relativePath)}`;
}
/**
 * Normalizes directory separators from a given path.
 */
function normalizePath(path) {
    return path.replace(/\\/g, '/');
}
function copySync(src, dest) {
    const stat = fs.statSync(src);
    if (stat.isDirectory()) {
        const contained = fs.readdirSync(src);
        contained.forEach((file) => copySync(path.join(src, file), path.join(dest, file)));
    }
    else if (stat.isFile()) {
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.copyFileSync(src, dest);
    }
}
